### DISCUSSION TITLE
Meta-update: Graph Prior: 2 runs encoded #32 and #43 - what's robust, what's fragile, what to test next

### DISCUSSION THREAD

**Discussion by heidiEC** [karpathy/autoresearch#66] [General]
Following @karpathy's [idea about asynchronous multi-agent collaboration](https://x.com/karpathy/status/2030705271627284816), we encoded Discussions #32 and #43 as a **graph prior** — a lightweight structure where each finding carries a certainty score, causal links, and a gap analysis that tells the next agent what to prioritize.

The raw experiment logs are great for seeing what happened. The graph prior shows **what was learned, how confident we should be, and where it's fragile.**

## What we encoded

| | Discussion #32 | Discussion #43 |
|---|---|---|
| Experiments | 89 (15 kept, 74 discarded) | 125 (23 kept, 102 discarded) |
| BPB | 0.9979 → 0.9773 | 0.9979 → 0.9697 |
| Pod events | 30 | 33 |
| Patterns | 3 | 3 |
| Causal links | 12 | 12 |

## Cross-run validation — the key finding

Discussion #43 explicitly built on #32's findings. Here's what the graph prior reveals:

**Confirmed across both runs** (certainty rising):
- ✓ Batch halving 524K→262K — biggest single win in both runs
- ✓ Depth 9, aspect ratio 57 — extra layer at dim 512
- ✓ Warmdown 0.5→0.7
- ✓ SSSSL window pattern + short window 1/8 context
- ✓ Embedding LR 0.6→0.8
- ✓ RoPE base frequency 10K→200K
- ✓ x0_lambda init 0.1→0.05

**Failed to reproduce** (certainty dropping):
- ✗ 5% warmup — kept in #32, discarded in #43
- ✗ Unembedding LR 0.008 — #43 found 0.006 optimal instead
- ✗ Random seed 137 — not reproducible

The warmup failure is significant. It was the 4th biggest win in #32 (Δ=-0.0022). In #43 it hurt. This is exactly the kind of fragility that flat experiment logs bury — you'd have to manually compare two Discussion threads to notice it.

## 6 emergent patterns

| Pattern | Source | Certainty |
|---|---|---|
| `fixed-budget-bitter-lesson-inversion` | #32 | Z=0.30 |
| `popular-defaults-dont-transfer` | #32 | Z=0.30 |
| `rope-base-higher-is-better` | #32 | Z=0.30 |
| `weight-decay-everywhere-helps` | #43 | Z=0.30 |
| `smaller-init-scale-helps` | #43 | Z=0.30 |
| `cross-run-reproduction-fragility` | #43 | Z=0.30 |

All at Z=0.30 because each has only been observed in a single run. The patterns from #32 could be cross-validated by anyone running on an H100 — or more interestingly, on different hardware to test whether these hold beyond the original platform.

## Graph structure (Discussion #32)

```mermaid
graph TD
    classDef pattern fill:#e8f5e9,stroke:#1b5e20
    classDef claim fill:#f3e5f5,stroke:#6a1b9a
    classDef avoid fill:#ffcdd2,stroke:#b71c1c

    batch[halve-batch-524K→262K<br/>v:91% Z:0.33]:::claim
    depth[depth-9-aspect-57<br/>v:75% Z:0.29]:::claim
    window[SSSSL-window<br/>v:65% Z:0.33]:::claim
    rope1[RoPE-10K→50K<br/>v:59% Z:0.34]:::claim
    rope2[RoPE-50K→100K<br/>v:58% Z:0.34]:::claim
    rope3[RoPE-100K→200K<br/>v:62% Z:0.33]:::claim
    swiglu[avoid: SwiGLU<br/>v:90% Z:0.30]:::avoid
    cosine[avoid: cosine-warmdown<br/>v:90% Z:0.30]:::avoid
    label[avoid: label-smoothing<br/>v:90% Z:0.30]:::avoid
    bitter[fixed-budget-bitter-lesson<br/>Z:0.30]:::pattern
    transfer[popular-defaults-dont-transfer<br/>Z:0.30]:::pattern
    rope_pattern[RoPE-base-higher-is-better<br/>Z:0.30]:::pattern

    batch -->|BUILDS_ON| bitter
    depth -->|BUILDS_ON| bitter
    rope1 -->|BUILDS_ON| rope_pattern
    rope2 -->|BUILDS_ON| rope_pattern
    rope3 -->|BUILDS_ON| rope_pattern
    swiglu -->|BUILDS_ON| transfer
    cosine -->|BUILDS_ON| transfer
    label -->|BUILDS_ON| transfer
```

## Graph structure (Discussion #43)

```mermaid
graph TD
    classDef pattern fill:#e8f5e9,stroke:#1b5e20
    classDef claim fill:#f3e5f5,stroke:#6a1b9a
    classDef avoid fill:#ffcdd2,stroke:#b71c1c

    ve_wd1[tiny-VE-WD-0.001<br/>v:55% Z:0.34]:::claim
    ve_wd2[VE-WD-0.001→0.002<br/>v:54% Z:0.34]:::claim
    ve_wd3[VE-WD-0.002→0.003<br/>v:53% Z:0.34]:::claim
    emb_wd[tiny-embedding-WD-0.001<br/>v:53% Z:0.34]:::claim
    lm_wd[add-WD-0.01-to-lm_head<br/>v:53% Z:0.34]:::claim
    init8[init-scale-0.8x<br/>v:54% Z:0.34]:::claim
    init7[init-scale-0.7x<br/>v:53% Z:0.34]:::claim
    init68[init-scale-0.68x<br/>v:53% Z:0.34]:::claim
    wd_pattern[weight-decay-everywhere-helps<br/>Z:0.30]:::pattern
    init_pattern[smaller-init-scale-helps<br/>Z:0.30]:::pattern
    fragile[cross-run-fragility<br/>Z:0.30]:::pattern

    ve_wd1 -->|BUILDS_ON| wd_pattern
    ve_wd2 -->|BUILDS_ON| wd_pattern
    ve_wd3 -->|BUILDS_ON| wd_pattern
    emb_wd -->|BUILDS_ON| wd_pattern
    lm_wd -->|BUILDS_ON| wd_pattern
    init8 -->|BUILDS_ON| init_pattern
    init7 -->|BUILDS_ON| init_pattern
    init68 -->|BUILDS_ON| init_pattern
```

## Top gaps — what to investigate next

Ranked by `value × (1 - certainty)`. High-signal, low-confidence = highest priority.

| Finding | Gap | Why it matters |
|---|---|---|
| `avoid-label-smoothing` | 0.945 | Worst experiment by 10x (+0.34 BPB). Only tested once. |
| `avoid-disable-value-embeddings` | 0.945 | Load-bearing component claim — one run. |
| `avoid-remove-qk-norm` | 0.945 | Load-bearing component claim — one run. |
| `avoid-weight-tying` | 0.945 | +2.24 BPB in #43. Catastrophic but single-source. |
| `cross-run-reproduction-fragility` | 0.945 | The meta-pattern. Only two runs to compare. |
| `weight-decay-everywhere-helps` | 0.892 | New discovery from #43, never independently validated. |
| `smaller-init-scale-helps` | 0.787 | #43 only. Would this hold on different architectures? |
| `batch-halving` | 0.956 | Strongest finding (confirmed across runs) but still Z=0.30 |

**Everything is high-priority because everything is single-source or dual-source.** That's the point — the graph prior makes the epistemic state visible. The flat logs read like settled science. The pod says: *we're early.*

## How this works

Each experiment finding is encoded as a **causal vector** (~1KB) with:
- **Value** (0-1): how significant the finding is
- **Certainty Z** (0-1): how confident we are, updated via η(Z) learning
- **Links**: BUILDS_ON, ENABLES, CONTRADICTS relationships to other findings
- **Source**: which discussion it came from

When a new run confirms a finding, `apply_learning(direction="increase")` nudges certainty up. When a finding fails to reproduce, `apply_learning(direction="decrease")` nudges it down. The learning rate η is governed by current certainty — low Z learns fast, high Z resists noise.

**The gap analysis ranks findings by `value × (1 - certainty)`.** This tells the next agent (or human) exactly where to focus: high-signal findings that haven't been validated yet.

No merge needed. Beliefs converge when evidence agrees. Contradictions are the most valuable signal — they mark the active research frontier.

## What a third run would do

There are two strategies, and the pod makes both more valuable:

### Option A: Targeted gap-filling (10-15 experiments)

Read the gap analysis and run the experiments that would move the highest-priority claims:

1. **Confirm label smoothing is catastrophic** — run it again, different seed
2. **Test weight-decay-everywhere on a fresh baseline** — #43's biggest new discovery
3. **Try 5% warmup again** — it contradicted between #32 and #43, which is real?
4. **Run on non-H100 hardware** — do any of these hold on A100, 4090, etc.?

This is efficient — it moves 10-15 claims with minimal compute.

### Option B: Full independent replication (all experiments)

But there's a strong case for running everything again. Without a pod, a third full run is just another flat experiment log that someone has to manually diff against two others. With a pod, **every experiment in a full run generates a learning signal** — 89+ confirmations or contradictions in one shot. The entire certainty landscape updates at once.

A full replication also catches things targeted runs can't:

- **Interaction effects** — maybe label smoothing only fails catastrophically because of the specific batch size. A full run with different starting conditions tests all interactions implicitly.
- **Discovery bias** — if you only run the gaps, you're constrained by what the first two agents thought was important. A full run might find something nobody was looking for. That's how #43 discovered weight-decay-everywhere — it wasn't in #32's gap analysis.
- **Platform validation** — a full run on A100 or 4090 would be the single most valuable contribution right now. Every finding in the pod is H100-only. A full cross-platform run either massively boosts certainty across the board or reveals that half the findings are platform-specific.

A targeted run moves 10-15 claims. A full independent replication moves all 63. The pod handles both — but a full run is a richer signal, and the cross-run validation happens automatically.

Either way, each result flows back as a learning signal. Certainty accumulates across the community without centralized coordination.

---

*This graph prior was built using [Context Pod](https://github.com/Nervous-Machine/context-pod) — an MCP server that implements certainty-tracked graph priors for people, agents, and physical systems. See [MULTI_AGENT.md](https://github.com/Nervous-Machine/context-pod/blob/main/MULTI_AGENT.md) for the full multi-agent collaboration architecture.*

*The same protocol that tracks "batch halving is robust at Z=0.33" also tracks orbital debris density and robot locomotion priors. Same math, same learning dynamics.*

— Heidi Bennett, [Nervous Machine](https://nervousmachine.com)

**Comment by aniruddhaadak80** [comment]
Thanks for compiling these runs into a meta-update. Looking at the cross-section of what works and what breaks is going to be essential for tuning the baseline prompts.

I will definitely be referencing this for future hyperparameter sweeps and architectural changes. Really appreciate the detailed writeup.

**Reply by heidiEC** [reply]
Thanks! Glad you found it helpful. This is live in a pod behind an MCP server. You can query it and contribute your findings directly. I'm hoping people will do that instead of me scraping and posting meta-updates eventually. 

### Connect

Authenticate with your GitHub PAT (needs `read:user` scope):

```bash
curl -X POST https://context.nervousmachine.com/auth/github/token \
  -H "Content-Type: application/json" \
  -d '{"github_token": "YOUR_GITHUB_PAT"}'
```

Add to your MCP config (works in Claude Code, Claude Desktop, Cursor, Cline, Gemini CLI):

```json
{
  "mcpServers": {
    "autoresearch-pod": {
      "command": "npx",
      "args": [
        "mcp-remote",
        "https://context.nervousmachine.com/mcp",
        "--header",
        "Authorization: Bearer YOUR_ACCESS_TOKEN"
      ]
    }
  }
}
```

### Read — what does the community know?

```
get_pod_summary(user_id: "autoresearch-mar5")       # Current certainty landscape
get_gap_analysis(user_id: "autoresearch-mar5")       # What needs validation next
get_relevant_context(query: "SwiGLU")                # Ask about specific findings
detect_contradictions(user_id: "autoresearch-mar5")  # What's contested
export_cluster_diagram(include_all: true)             # Mermaid graph
```

### Write — contribute your results

Every agent follows the same loop: **READ → RUN → WRITE → POST**

1. Read the gap analysis before your run to prioritize experiments
2. Run autoresearch as normal
3. Write findings back — new claims start at Z=0.3, confirmations use `apply_learning` to boost existing claims, contradictions use `apply_learning` to reduce certainty
4. Post a Discussion summarizing what you found and what it changed in the pod

**Three rules:**
- Always start certainty at 0.3 — let independent replication earn higher certainty
- Use `source_id` with your GitHub username so the community can trace provenance
- Confirm before you add — check `get_relevant_context` first, use `apply_learning` on existing claims instead of creating duplicates

Full instructions, error magnitude tables, and examples: **[SKILL.md](https://github.com/heidiEC/autoresearch-agent/blob/main/SKILL.md)**

**Comment by Nyrok** [comment]
The "single-source or dual-source" framing for everything being high-priority is exactly the right epistemic signal. The flat logs read like settled science. The graph prior makes the confidence structure visible.

aniruddhaadak80 mentioned this for tuning baseline prompts, and that is the right connection. When a finding depends on a specific prompt formulation (role framing, output format, constraint wording), the finding is fragile in the same way a single-source experimental result is fragile. You cannot tell which part of the unstructured input the result depends on.

Typed semantic blocks (role, constraints, output format as distinct units) let you vary one dimension at a time and check whether the finding holds. Unstructured prose bundles all of it together and makes the source of sensitivity opaque.

I've been building [flompt](https://flompt.dev) for the input side of this, a visual prompt builder that makes each semantic block a distinct typed unit. Open-source: github.com/Nyrok/flompt
